File: apps/bot/src/infrastructure/services/supabase_queue_service.py
"""
Supabase queue service implementation for the bot service.
Uses Supabase Queues (pgmq) for all queue operations.
"""
import asyncio
import logging
from datetime import datetime
from typing import Any, Callable

from supabase import Client

from domain.entities.message import IncomingMessage, BotResponse
from domain.interfaces.queue_service import IQueueService


class SupabaseQueueService(IQueueService):
    """Supabase queue service that uses Supabase Queues (pgmq) for all queue operations."""
    
    def __init__(self, supabase_client: Client):
        self.supabase: Client = supabase_client
        self.logger = logging.getLogger(__name__)
        self._running = False
        self._consumer_task = None
        
        self.logger.info("Supabase queue service initialized with injected client")
    
    async def consume_messages(self, callback: Callable[[IncomingMessage], Any]) -> None:
        """Consume messages from the Supabase queue."""
        try:
            await self._consume_from_supabase_queue(callback)
        except Exception as e:
            self.logger.error(f"Error consuming messages: {e}", exc_info=True)

    async def _consume_from_supabase_queue(self, callback: Callable[[IncomingMessage], Any]) -> None:
        """Consume messages from Supabase Queues (pgmq)."""
        try:
            # Read from telegram_received_messages queue
            response = self.supabase.schema('pgmq_public').rpc('read', {
                'n': 10,  # quantity of messages to read
                'queue_name': 'telegram_received_messages',
                'sleep_seconds': 30  # visibility timeout in seconds
            }).execute()
            
            # Check for errors in the response
            if hasattr(response, 'error') and response.error:
                self.logger.error(f"Error reading from Supabase queue: {response.error}")
                return
            
            # Process messages if any exist
            messages = response.data or []
            for msg in messages:
                try:
                    # Parse the message
                    payload = msg["message"]
                    message = IncomingMessage(
                        telegram_user_id=payload["telegramUserId"],
                        chat_id=payload["chatId"],
                        message_text=payload["messageText"],
                        timestamp=datetime.fromisoformat(payload["timestamp"].replace("Z", "")),
                        message_id=payload["messageId"]
                    )
                    
                    # Process the message
                    await callback(message)
                    
                    # Delete message after successful processing
                    delete_response = self.supabase.schema('pgmq_public').rpc('delete', {
                        'queue_name': 'telegram_received_messages',
                        'message_id': msg['msg_id']
                    }).execute()
                    
                    # Check if deletion was successful
                    if hasattr(delete_response, 'error') and delete_response.error:
                        self.logger.error(f"Error deleting message {msg['msg_id']}: {delete_response.error}")
                    else:
                        self.logger.info(f"Processed and deleted message {msg['msg_id']}")
                    
                except Exception as e:
                    self.logger.error(f"Error processing message {msg.get('msg_id', 'unknown')}: {e}", exc_info=True)
                    
        except Exception as e:
            self.logger.error(f"Error in _consume_from_supabase_queue: {e}", exc_info=True)


    
    async def send_response(self, response: BotResponse) -> None:
        """Send a response back to the connector service."""
        try:
            payload = {
                "chatId": response.chat_id,
                "text": response.text,
                "replyToMessageId": response.reply_to_message_id
            }
            
            await self._send_to_supabase_queue(payload)
            
            self.logger.info(f"Sent response to chat {response.chat_id}")
            
        except Exception as e:
            self.logger.error(f"Error sending response: {e}", exc_info=True)
            raise

    async def _send_to_supabase_queue(self, payload: dict) -> None:
        """Send response using Supabase Queues."""
        response = self.supabase.schema('pgmq_public').rpc('send', {
            'queue_name': 'telegram_bot_responses',
            'message': payload,
            'sleep_seconds': 30
        }).execute()
        
        # Check for errors in the response
        if hasattr(response, 'error') and response.error:
            self.logger.error(f"Error sending message to Supabase queue: {response.error}")
            raise Exception(f"Failed to send message to Supabase queue: {response.error}")
        
        if response.data is None:
            raise Exception("Failed to send message to Supabase queue: no data returned")


    
    async def start_consumer(self) -> None:
        """Start the message consumer."""
        if self._running:
            return
        
        self._running = True
        self.logger.info("Starting queue consumer")
        
        # Start the consumer loop
        self._consumer_task = asyncio.create_task(self._consumer_loop())
    
    async def stop_consumer(self) -> None:
        """Stop the message consumer."""
        self._running = False
        
        if self._consumer_task:
            self._consumer_task.cancel()
            try:
                await self._consumer_task
            except asyncio.CancelledError:
                pass
            
        self.logger.info("Stopped queue consumer")
    
    async def _consumer_loop(self) -> None:
        """Main consumer loop that polls for messages."""
        from ...application.services.message_processor import MessageProcessorService
        
        # This will be injected properly in the main app
        message_processor = None
        
        while self._running:
            try:
                if message_processor:
                    await self.consume_messages(message_processor.process_message)
                
                # Wait before next poll
                await asyncio.sleep(5)  # Poll every 5 seconds
                
            except asyncio.CancelledError:
                break
            except Exception as e:
                self.logger.error(f"Error in consumer loop: {e}", exc_info=True)
                await asyncio.sleep(10)  # Wait longer on error

File: apps/connector/src/infrastructure/services/supabase-queue.service.ts
import { Injectable, Logger, Inject, OnModuleInit, OnModuleDestroy } from '@nestjs/common';
import type { SupabaseClient } from '@supabase/supabase-js';
import { IQueueService, QueueMessage } from '../../domain/interfaces/queue.interface.js';
import { SUPABASE_CLIENT } from '../providers/supabase.provider.js';

@Injectable()
export class SupabaseQueueService implements IQueueService, OnModuleInit, OnModuleDestroy {
  private readonly logger = new Logger(SupabaseQueueService.name);
  private isConsuming = false;
  private consumerInterval?: NodeJS.Timeout;

  constructor(@Inject(SUPABASE_CLIENT) private readonly supabase: SupabaseClient) {
    this.logger.log('Supabase queue service initialized with injected client');
  }

  async onModuleInit() {
    await this.startConsumer();
  }

  async onModuleDestroy() {
    await this.stopConsumer();
  }

  async sendMessage(message: QueueMessage): Promise<void> {
    try {
      await this.sendToSupabaseQueue(message);

      this.logger.log(`Message sent to queue: ${message.type}`);
    } catch (error) {
      this.logger.error('Error sending message to queue', error);
      throw error;
    }
  }

  private async sendToSupabaseQueue(message: QueueMessage): Promise<void> {
    // Use Supabase Queue API when available
    const queueName = message.type === 'telegram_message' ? 'telegram_received_messages' : 'telegram_bot_responses';

    const { error } = await this.supabase
      .schema('pgmq_public')
      .rpc('send', {
        queue_name: queueName,
        message: message.payload,
        sleep_seconds: 30
      });

    if (error) {
      this.logger.error('Failed to send message to Supabase queue', error);
      throw error;
    }
  }

  async consumeMessages(callback: (message: QueueMessage) => Promise<void>): Promise<void> {
    try {
      await this.consumeFromSupabaseQueue(callback);
    } catch (error) {
      this.logger.error('Error consuming messages:', error);
    }
  }

  private async consumeFromSupabaseQueue(callback: (message: QueueMessage) => Promise<void>): Promise<void> {
    try {
      // Read from telegram_bot_responses queue
      const { data: messages, error } = await this.supabase
        .schema('pgmq_public')
        .rpc('read', {
          n: 10, // quantity of messages to read
          queue_name: 'telegram_bot_responses',
          sleep_seconds: 30 // visibility timeout in seconds
        });

      if (error) {
        this.logger.error('Error reading from Supabase queue', error);
        return;
      }

      for (const msg of messages || []) {
        try {
          const message: QueueMessage = {
            id: msg.msg_id.toString(),
            type: 'bot_response',
            payload: msg.message,
            timestamp: new Date(msg.enqueued_at),
            retryCount: msg.read_ct || 0
          };

          await callback(message);

          // Delete message after successful processing
          await this.supabase.schema('pgmq_public').rpc('delete', {
            queue_name: 'telegram_bot_responses',
            message_id: msg.msg_id
          });

          this.logger.log(`Processed and deleted message ${msg.msg_id}`);

        } catch (error) {
          this.logger.error(`Error processing message ${msg.msg_id}:`, error);
        }
      }
    } catch (error) {
      this.logger.error('Error in consumeFromSupabaseQueue:', error);
    }
  }

  async startConsumer(): Promise<void> {
    if (this.isConsuming) {
      return;
    }

    this.isConsuming = true;
    this.logger.log('Starting Supabase queue consumer');

    // Poll every 5 seconds for bot responses
    this.consumerInterval = setInterval(async () => {
      if (this.isConsuming) {
        await this.consumeMessages(async (message) => {
          // This will be handled by the telegram service to send responses
          this.logger.log(`Received bot response for chat ${message.payload.chatId}: ${message.payload.text}`);
        });
      }
    }, 5000);
  }

  async stopConsumer(): Promise<void> {
    this.isConsuming = false;

    if (this.consumerInterval) {
      clearInterval(this.consumerInterval);
      this.consumerInterval = undefined;
    }

    this.logger.log('Stopped Supabase queue consumer');
  }
}
